virtio-device/src/virtio_config.rs

```diff
@@ -130,6 +182,10 @@ where
             1 => ((features << 32) >> 32) + (v << 32),
             // Accessing an unknown page has no effect.
             _ => features,
+        };
+
+        if page == 0 {
```

Comment:
Why only on page 0?

Is it fine if for some strange reason the device negotiates only features on page 1 to not call `negotiate_driver_features()`?

Reply:
> Why only on page 0?
> 
> Is it fine if for some strange reason the device negotiates only features on page 1 to not call `negotiate_driver_features()`?


In accordance with Linux, the frontend driver negotiates the device features recognized and enabled by the driver by first selecting page 1, performing a write operation, and then selecting page 0 to perform a subsequent write (check the `vm_finalize_features` function).

However, if for some unexpected reason, the driver only negotiates features on page 1, how can we detect this scenario? Would it be feasible to negotiate the driver features and protocol features (e.g., vhost-user) with the backend device twice?

It seems more logical to negotiate the features only once, ensuring that all driver features have been written and are fully prepared for negotiation (by verifying page 0). 

Additionally, we could delegate the check for page 0 to the VMM itself by invoking the `driver_features_select` method to determine which page is currently selected. However, it seems more efficient to integrate this logic directly into the `negotiate_driver_features` function.

Reply:
> > Why only on page 0?
> > Is it fine if for some strange reason the device negotiates only features on page 1 to not call `negotiate_driver_features()`?
> 
> In accordance with Linux, the frontend driver negotiates the device features recognized and enabled by the driver by first selecting page 1, performing a write operation, and then selecting page 0 to perform a subsequent write (check the `vm_finalize_features` function).

Is it defined by the spec, or it is an implementation detail?

> 
> However, if for some unexpected reason, the driver only negotiates features on page 1, how can we detect this scenario? 

If the spec defined the behaviour you mentioned, it is fine. Just add a comment on the code with that.

>Would it be feasible to negotiate the driver features and protocol features (e.g., vhost-user) with the backend device twice?

Maybe it's supported, but not ideally.

> 
> It seems more logical to negotiate the features only once, ensuring that all driver features have been written and are fully prepared for negotiation (by verifying page 0).

Yeah, I see.

> 
> Additionally, we could delegate the check for page 0 to the VMM itself by invoking the `driver_features_select` method to determine which page is currently selected. However, it seems more efficient to integrate this logic directly into the `negotiate_driver_features` function.

Agree.

Reply:
> Is it defined by the spec, or it is an implementation detail?

According to the [VirtIO specification](https://docs.oasis-open.org/virtio/virtio/v1.3/csd01/virtio-v1.3-csd01.html), I was unable to find specific guidance regarding how the driver should negotiate the features it supports. However, upon reviewing other VirtIO driver implementations, such as the [VirtIO guest drivers in Rust](https://crates.io/crates/virtio-drivers), I discovered that they implement the negotiation process in the reverse order compared to Linux. Specifically, they first select page 0 to address the lower 32 bits of features before moving to the higher 32 bits on page 1, as seen [here](https://github.com/rcore-os/virtio-drivers/blob/master/src/transport/mmio.rs#L338).
However, in [Linux](https://github.com/torvalds/linux/blob/master/drivers/virtio/virtio_mmio.c#L124), this happens in reverse order.
So, this indicates that the negotiation order can be implementation-specific.

A potential solution could be to verify if the feature negotiation is complete by checking the `FEATURES_OK` bit in the `set_device_status` method. 
If the bit is set, we can proceed to invoke the hook-up. 
This approach would provide abstraction from the guest side, allowing it to write to either page 0 or page 1, while ensuring that the hook-up is implemented only after all features have been successfully negotiated between the frontend and backend. 
What are your thoughts on this approach? If you agree, I can do that.

Reply:
> > Is it defined by the spec, or it is an implementation detail?
> 
> According to the [VirtIO specification](https://docs.oasis-open.org/virtio/virtio/v1.3/csd01/virtio-v1.3-csd01.html), I was unable to find specific guidance regarding how the driver should negotiate the features it supports. However, upon reviewing other VirtIO driver implementations, such as the [VirtIO guest drivers in Rust](https://crates.io/crates/virtio-drivers), I discovered that they implement the negotiation process in reverse order compared to [Linux](https://github.com/torvalds/linux/blob/master/drivers/virtio/virtio_mmio.c#L124). Specifically, they first select page 0 to address the lower 32 bits of features before moving to the higher 32 bits on page 1, as seen [here](https://github.com/rcore-os/virtio-drivers/blob/master/src/transport/mmio.rs#L338). So, this indicates that the negotiation order can be implementation-specific.

That was exactly my concern!

> 
> A potential solution could be to verify if the feature negotiation is complete by checking the `FEATURES_OK` bit in the `set_device_status` method. If the bit is set, we can proceed to invoke the hook-up. This approach would provide abstraction from the guest side, allowing it to write to either page 0 or page 1, while ensuring that the hook-up is implemented only after all features have been successfully negotiated between the frontend and backend. What are your thoughts on this approach? If you agree, I can do that.

Yeah, I completely agree!

Reply:
> > A potential solution could be to verify if the feature negotiation is complete by checking the `FEATURES_OK` bit in the `set_device_status` method. If the bit is set, we can proceed to invoke the hook-up. This approach would provide abstraction from the guest side, allowing it to write to either page 0 or page 1, while ensuring that the hook-up is implemented only after all features have been successfully negotiated between the frontend and backend. What are your thoughts on this approach? If you agree, I can do that.
> 
> Yeah, I completely agree!

Done [here](https://github.com/rust-vmm/vm-virtio/pull/291/commits/39eeae40e89870872786841977b2731d049343bb)

Reply:
> > > A potential solution could be to verify if the feature negotiation is complete by checking the `FEATURES_OK` bit in the `set_device_status` method. If the bit is set, we can proceed to invoke the hook-up. This approach would provide abstraction from the guest side, allowing it to write to either page 0 or page 1, while ensuring that the hook-up is implemented only after all features have been successfully negotiated between the frontend and backend. What are your thoughts on this approach? If you agree, I can do that.
> > 
> > 
> > Yeah, I completely agree!
> 
> Done [here](https://github.com/rust-vmm/vm-virtio/pull/291/commits/39eeae40e89870872786841977b2731d049343bb)

Cool, LGTM!
I think you can merge that change with the first commit and I'll approve it ;-)

Reply:
> > > > A potential solution could be to verify if the feature negotiation is complete by checking the `FEATURES_OK` bit in the `set_device_status` method. If the bit is set, we can proceed to invoke the hook-up. This approach would provide abstraction from the guest side, allowing it to write to either page 0 or page 1, while ensuring that the hook-up is implemented only after all features have been successfully negotiated between the frontend and backend. What are your thoughts on this approach? If you agree, I can do that.
> > > 
> > > 
> > > Yeah, I completely agree!
> > 
> > 
> > Done [here](https://github.com/rust-vmm/vm-virtio/pull/291/commits/39eeae40e89870872786841977b2731d049343bb)
> 
> Cool, LGTM! I think you can merge that change with the first commit and I'll approve it ;-)

Done :)